### PR TITLE
Support dleks in select map (By D1GQ )

### PR DESCRIPTION
**Pull request by Tommy-XL**
Sad

image

### CODE REVIEW

**Comment by Ultradragon005** [comment]
Object Reference Not Set To An Instance Of An Object 😢😭😭😭😭

**Comment by Tommy-XL** [comment]
I'll have to postpone this for 2.1.0 because I noticed problems that I don't want to add to the release version

(Anyway, if AU is updated we will have to release 2.0.1 and there we can add it if it is completed)

**Comment by D1GQ** [comment]
@Tommy-XL Try this
```
        [HarmonyPatch(nameof(GameOptionsMapPicker.FixedUpdate))]
        [HarmonyPrefix]
        public static bool Prefix_FixedUpdate(GameOptionsMapPicker __instance)
        {
            if (DleksButton != null)
            {
                if (__instance.selectedButton == DleksButton)
                {
                    SetDleks = true;
                }
                else
                {
                    SetDleks = false;
                }
            }

            if (__instance.selectedMapId == 3)
                return false;

            return true;
        }
```

**Comment by Tommy-XL** [comment]
The problems that I noticed were:

1. When a lobby is created and dleks is selected, a null error appears
2. When you click or switch maps several times, the map icons are reset and a null error appears, and for this I added a “try catch”
3. When the host presses "start" the buttons in the settings are reset until the host selects the map again

Maybe there are some other problems that I did not notice and therefore did not add this change to version 2.0.0

**Comment by D1GQ** [comment]
@Tommy-XL This should fix all the issues, It was a simple check that was missing for __instance.selectedButton.Button.SelectButton(false);

```
    [HarmonyPatch(typeof(GameOptionsMapPicker))]
    public static class GameOptionsMapPickerPatch
    {
        [HarmonyPatch(nameof(GameOptionsMapPicker.Initialize))]
        [HarmonyPostfix]
        public static void Postfix_Initialize(GameOptionsMapPicker __instance)
        {
            int DleksPos = 3;

            MapSelectButton[] AllMapButton = __instance.transform.GetComponentsInChildren<MapSelectButton>();

            if (AllMapButton != null)
            {
                GameObject dlekS_ehT = UnityEngine.Object.Instantiate(AllMapButton[0].gameObject, __instance.transform);
                dlekS_ehT.transform.position = AllMapButton[DleksPos].transform.position;
                dlekS_ehT.transform.SetSiblingIndex(DleksPos + 2);
                MapSelectButton dlekS_ehT_MapButton = dlekS_ehT.GetComponent<MapSelectButton>();
                DleksButton = dlekS_ehT_MapButton;
                dlekS_ehT_MapButton.MapIcon.transform.localScale = new Vector3(-1f, 1f, 1f);
                dlekS_ehT_MapButton.Button.OnClick.RemoveAllListeners();
                dlekS_ehT_MapButton.Button.OnClick.AddListener((Action)(() =>
                {
                    __instance.SelectMap(__instance.AllMapIcons[0]);

                    if (__instance.selectedButton)
                    {
                        __instance.selectedButton.Button.SelectButton(false);
                    }
                    __instance.selectedButton = dlekS_ehT_MapButton;
                    __instance.selectedButton.Button.SelectButton(true);
                    __instance.selectedMapId = 3;

                    if (GameStates.IsNormalGame)
                        Main.NormalOptions.MapId = 0;
                    else if (GameStates.IsHideNSeek)
                        Main.HideNSeekOptions.MapId = 0;

                    Main.NormalOptions.MapId = 0;

                    __instance.MapImage.sprite = Utils.LoadSprite($"TOHE.Resources.Images.DleksBanner.png", 100f);
                    __instance.MapName.sprite = Utils.LoadSprite($"TOHE.Resources.Images.DleksBanner-Wordart.png", 100f);
                }));

                for (int i = DleksPos; i < AllMapButton.Length; i++)
                {
                    AllMapButton[i].transform.localPosition += new Vector3(0.625f, 0f, 0f);
                }

                if (DleksButton != null)
                {
                    if (SetDleks == true)
                    {
                        if (__instance.selectedButton)
                        {
                            __instance.selectedButton.Button.SelectButton(false);
                        }
                        DleksButton.Button.SelectButton(true);
                        __instance.selectedButton = DleksButton;
                        __instance.selectedMapId = 3;

                        __instance.MapImage.sprite = Utils.LoadSprite($"TOHE.Resources.Images.DleksBanner.png", 100f);
                        __instance.MapName.sprite = Utils.LoadSprite($"TOHE.Resources.Images.DleksBanner-Wordart.png", 100f);
                    }
                    else
                    {
                        DleksButton.Button.SelectButton(false);
                    }
                }
            }
        }
```

```
        [HarmonyPatch(nameof(GameOptionsMapPicker.FixedUpdate))]
        [HarmonyPrefix]
        public static bool Prefix_FixedUpdate(GameOptionsMapPicker __instance)
        {
            if (DleksButton != null)
            {
                if (__instance.selectedMapId == 3)
                {
                    SetDleks = true;
                }
                else
                {
                    SetDleks = false;
                }
            }

            if (__instance.selectedMapId == 3)
                return false;

            return true;
        }
```

**Comment by Tommy-XL** [comment]
Seems done

File: Patches/DleksPatch.cs [hunk 1]
```diff
@@ -45,8 +45,27 @@ public static void Postfix_AllMapIcons(GameStartManager __instance)
...
+                CreateOptionsPickerPatch.SetDleks = true;
+        }
+        else if (GameStates.IsHideNSeek && Main.HideNSeekOptions.MapId == 3)
+        {
+            Main.HideNSeekOptions.MapId = 0;
+            __instance.UpdateMapImage(MapNames.Skeld);
+
+            if (!Options.RandomMapsMode.GetBool())
+                CreateOptionsPickerPatch.SetDleks = true;
+        }
```
Comment (Ultradragon005) [inline Patches/DleksPatch.cs:63]: isn't this the exact same code jsut do a || statement in the thing above 😆

Comment (Ultradragon005) [reply Patches/DleksPatch.cs:63]: nvm I am blind

Comment (Tommy-XL) [reply Patches/DleksPatch.cs:63]: ?

Comment (Ultradragon005) [reply Patches/DleksPatch.cs:63]: forget about it I just didn't see something

**Pull request merged**